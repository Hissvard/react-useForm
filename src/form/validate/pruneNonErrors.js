const pruneArray = errors => {
  let hasError = false
  const results = errors.reduce((acc, value) => {
    const children = pruneObject(value)
    if (children && Object.keys(children).length > 0) {
      hasError = true
    }

    acc.push(children)
    return acc
  }, [])

  return hasError ? results : []
}

const pruneObject = errors => {
  if (!errors) return errors

  return Object.entries(errors).reduce((acc, [key, value]) => {
    if (Array.isArray(value)) {
      const children = pruneArray(value)
      if (children.length > 0) {
        acc[key] = children
      }
      if (value.error) {
        if (!acc[key]) acc[key] = []
        acc[key].error = value.error
      }
      return acc
    }

    if (typeof value === 'object') {
      const children = pruneObject(value)
      if (children && Object.keys(children).length > 0) {
        acc[key] = children
      }
    } else if (value) {
      acc[key] = value
    }

    return acc
  }, {})
}

export const pruneNonErrors = errors => {
  return pruneObject(errors)
}
