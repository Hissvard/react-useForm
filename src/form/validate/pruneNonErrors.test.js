import { pruneNonErrors } from './pruneNonErrors'

let errors
beforeEach(() => {

})

test('prunes first level', () => {
  errors = {
    nickname: '',
    colors: [],
  }

  expect(pruneNonErrors(errors)).toMatchSnapshot()
})

test('prunes more levels', () => {
  errors = {
    nickname: '',
    colors: [{
      color: 'color',
      shades: [
        { shade: '' },
      ],
    }],
  }

  expect(pruneNonErrors(errors)).toMatchSnapshot()
})

test('prunes nested object', () => {
  errors = {
    name: {
      first: '',
      middle: 'finger',
      last: 'last',
    },
    colors: [{
      color: 'color',
      shades: [
        { shade: '' },
      ],
    }],
  }

  expect(pruneNonErrors(errors)).toMatchSnapshot()
})

test('prunes all of nested objects', () => {
  errors = {
    name: {
      first: '',
      middle: '',
      last: '',
    },
    colors: [
      undefined,
      {
        color: '',
        shades: [
          null,
          { shade: '' },
          undefined,
        ],
      }],
  }

  expect(pruneNonErrors(errors)).toMatchSnapshot()
})

test('prunes nested array', () => {
  errors = {
    nickname: '',
    colors: [{
      shades: [
        undefined,
        { shade: 'shade' },
      ],
    }],
  }

  expect(pruneNonErrors(errors)).toMatchSnapshot()
})

test('prunes array at middle index', () => {
  errors = {
    nickname: '',
    colors: [
      undefined,
      { color: 'color' },
      undefined,
    ],
  }

  expect(pruneNonErrors(errors)).toMatchSnapshot()
})

test('leaves error on array', () => {
  errors = {
    nickname: '',
    colors: [
      undefined,
      { color: '' },
      undefined,
    ],
  }
  errors.colors.error = 'colors'

  expect(pruneNonErrors(errors)).toMatchSnapshot()
  expect(errors.colors.error).toEqual('colors')
})

test('leaves error on array with errors', () => {
  errors = {
    nickname: '',
    colors: [
      undefined,
      { color: 'color' },
      undefined,
    ],
  }
  errors.colors.error = 'colors'

  expect(pruneNonErrors(errors)).toMatchSnapshot()
  expect(errors.colors.error).toEqual('colors')
})
