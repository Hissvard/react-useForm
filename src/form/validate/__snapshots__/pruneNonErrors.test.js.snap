// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`leaves error on array 1`] = `
Object {
  "colors": Array [],
}
`;

exports[`leaves error on array with errors 1`] = `
Object {
  "colors": Array [
    undefined,
    Object {
      "color": "color",
    },
    undefined,
  ],
}
`;

exports[`prunes all of nested objects 1`] = `Object {}`;

exports[`prunes array at middle index 1`] = `
Object {
  "colors": Array [
    undefined,
    Object {
      "color": "color",
    },
    undefined,
  ],
}
`;

exports[`prunes first level 1`] = `Object {}`;

exports[`prunes more levels 1`] = `
Object {
  "colors": Array [
    Object {
      "color": "color",
    },
  ],
}
`;

exports[`prunes nested array 1`] = `
Object {
  "colors": Array [
    Object {
      "shades": Array [
        undefined,
        Object {
          "shade": "shade",
        },
      ],
    },
  ],
}
`;

exports[`prunes nested object 1`] = `
Object {
  "colors": Array [
    Object {
      "color": "color",
    },
  ],
  "name": Object {
    "last": "last",
    "middle": "finger",
  },
}
`;
